pkg/jobstore/boltdb/store.go

```diff
@@ -690,11 +699,94 @@ func (b *BoltJobStore) CreateJob(ctx context.Context, job models.Job, event mode
 	if err != nil {
 		return err
 	}
-	return b.database.Update(func(tx *bolt.Tx) (err error) {
+	return b.update(ctx, func(tx *bolt.Tx) (err error) {
 		return b.createJob(tx, job, event)
 	})
 }
 
+// update is a helper function that will update the job in the store
+// it accepts a context, an update function and creates a new transaction to
+// perform the update if no transaction is provided in the context
+func (b *BoltJobStore) update(ctx context.Context, update func(tx *bolt.Tx) error) error {
+	var err error
+	var tx *bolt.Tx
+	var managed bool
+
+	// if ctx has a transaction value, then we can use that transaction
+	// and mark it as managed so that we don't try to commit it or rollback in this function
+	var txFound bool
+	tx, txFound = txFromContext(ctx)
+	if txFound {
+		if !tx.Writable() {
+			return fmt.Errorf("readonly transaction provided in context for update operation")
+		}
+		managed = true
+	} else {
+		tx, err = b.database.Begin(true)
+		if err != nil {
+			return err
+		}
+	}
+
+	// always rollback the transaction if we did create it and we're returning an error
+	defer func() {
+		if !managed && err != nil {
+			_ = tx.Rollback()
```

Comment:
should we return this error to the caller?

Reply:
similar to https://github.com/bacalhau-project/bacalhau/pull/3996#discussion_r1604832916